Readme.md
## GDLogger

[![Platforms](https://img.shields.io/cocoapods/p/GDLogger.svg)](https://cocoapods.org/pods/GDLogger)
[![License](https://img.shields.io/cocoapods/l/GDLogger.svg)](https://raw.githubusercontent.com/genedelisa/GDLogger/master/LICENSE)

[![Swift Package Manager](https://img.shields.io/badge/Swift%20Package%20Manager-compatible-brightgreen.svg)](https://github.com/apple/swift-package-manager)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![CocoaPods compatible](https://img.shields.io/cocoapods/v/GDLogger.svg)](https://cocoapods.org/pods/GDLogger)

[![Travis](https://img.shields.io/travis/genedelisa/GDLogger/master.svg)](https://travis-ci.org/genedelisa/GDLogger/branches)

[![Swift 4](https://img.shields.io/badge/swift4-compatible-4BC51D.svg?style=flat")](https://developer.apple.com/swift)

[![Platforms iOS](https://img.shields.io/badge/Platforms-iOS-lightgray.svg?style=flat)](https://swift.org/)
[![Platforms macOS](https://img.shields.io/badge/Platforms-macOS-lightgray.svg?style=flat)](https://swift.org/)
[![Platforms tvOS](https://img.shields.io/badge/Platforms-tvOS-lightgray.svg?style=flat)](https://swift.org/)
[![Platforms watchOS](https://img.shields.io/badge/Platforms-watchOS-lightgray.svg?style=flat)](https://swift.org/)

[![Sponsors](https://img.shields.io/badge/Sponsors-Rockhopper%20Technologies-orange.svg?style=flat)](http://www.rockhoppertech.com/)
[![Twitter @GeneDeLisaDev](https://img.shields.io/twitter/follow/GeneDeLisaDev.svg?style=social)](https://twitter.com/GeneDeLisaDev)



Yet another logger for Swift.
This uses Apple's Unified Logging `OSLog` - and `NSLog` as a fallback, so it's essentially a façade.
I think this makes `OSLog` easier to use in Swift since you can use string interpolation.



- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [License](#license)

## Requirements

- iOS 10.0+ / Mac OS X 10.12+ / tvOS 10.0+ / watchOS 3.0+
- Xcode 9.0+

## Installation

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects. You can install it with the following command:

```bash
$ gem install cocoapods
```


To integrate GDLogger into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '10.0'
use_frameworks!

pod 'GDLogger', '~> 0.0.1'
```

Then, run the following command:

```bash
$ pod install
```

### Carthage

[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that automates the process of adding frameworks to your Cocoa application.

You can install Carthage with [Homebrew](http://brew.sh/) using the following command:

```bash
$ brew update
$ brew install carthage
```

To integrate GDLogger into your Xcode project using Carthage, specify it in your `Cartfile`:

```ogdl
github "GDLogger/GDLogger" ~> 0.0.1
```
### Swift Package Manager

To use GDLogger as a [Swift Package Manager](https://swift.org/package-manager/) package just add the following in your Package.swift file.

``` swift
import PackageDescription

let package = Package(
    name: "HelloGDLogger",
    dependencies: [
        .Package(url: "https://github.com/genedelisa/GDLogger.git", "0.0.1")
    ]
)
```

### Manually

If you prefer not to use either of the aforementioned dependency managers, you can integrate GDLogger into your project manually.

#### Git Submodules

- Open up Terminal, `cd` into your top-level project directory, and run the following command "if" your project is not initialized as a git repository:

```bash
$ git init
```

- Add GDLogger as a git [submodule](http://git-scm.com/docs/git-submodule) by running the following command:

```bash
$ git submodule add https://github.com/genedelisa/GDLogger.git
$ git submodule update --init --recursive
```

- Open the new `GDLogger` folder, and drag the `GDLogger.xcodeproj` into the Project Navigator of your application's Xcode project.

    > It should appear nested underneath your application's blue project icon. Whether it is above or below all the other Xcode groups does not matter.

- Select the `GDLogger.xcodeproj` in the Project Navigator and verify the deployment target matches that of your application target.
- Next, select your application project in the Project Navigator (blue project icon) to navigate to the target configuration window and select the application target under the "Targets" heading in the sidebar.
- In the tab bar at the top of that window, open the "General" panel.
- Click on the `+` button under the "Embedded Binaries" section.
- You will see two different `GDLogger.xcodeproj` folders each with two different versions of the `GDLogger.framework` nested inside a `Products` folder.

    > It does not matter which `Products` folder you choose from.

- Select the `GDLogger.framework`.

- And that's it!

> The `GDLogger.framework` is automagically added as a target dependency, linked framework and embedded framework in a copy files build phase which is all you need to build on the simulator and a device.

#### Embeded Binaries

- Download the latest release from https://github.com/genedelisa/GDLogger/releases
- Next, select your application project in the Project Navigator (blue project icon) to navigate to the target configuration window and select the application target under the "Targets" heading in the sidebar.
- In the tab bar at the top of that window, open the "General" panel.
- Click on the `+` button under the "Embedded Binaries" section.
- Add the downloaded `GDLogger.framework`.
- And that's it!

### Other logging options.

[CocoaLumberjack](https://github.com/CocoaLumberjack/CocoaLumberjack)

[XCGLogger](https://github.com/DaveWoodCom/XCGLogger)


## Usage

[Apple's Unified Logging](https://developer.apple.com/documentation/os/logging)

To set the logging level for a subsystem.
```bash
$ sudo log config --mode "level:debug" --subsystem com.your_company.your_subsystem_name
```
For example, from the Terminal (or iterm), set the debug level for your subsystem. Then you can stream the output.
```bash

$ sudo log config --mode "level:debug" --subsystem com.rockhoppertech.SlowItDown

$ log stream --predicate 'subsystem == "com.rockhoppertech.SlowItDown"' --info --debug
```
To check the current logging level.
```bash
$ sudo log config --status
```

Use the system console (and/or Xcode's console) to read the logging messages.

You can filter by your subsystem in the Console app.

Sample Console output:

![Console](screenshots/consoleLogging.png "Console")


The message emojis can be redefined per level.

```Swift
var log = GDLogger()
log.debugPrefix = "♫ "
log.debugPostfix = "  ♫"
```

Here is some sample output.

```bash
2017-12-18 15:54:03.611598-0500 SlowItDown[31758:2620634] [general] 😺😺😺 authorized ☞ checkMediaLibraryPermission() 🗄MediaLibraryController.swift➸121 😺😺😺
```

Other options for creating loggers.

```Swift
var log2 = GDLogger(category: "my category")
log2.debug("logging to my category string")

public enum MyCategories: String {
    case mycat
}
var log3 = GDLogger(category: MyCategories.mycat.rawValue)
log3.debug("logging to my category enum")

var log4 = GDLogger(.controller, category: .general)
log4.debug("logging to controler subsystem using general category")
````

## License

GDLogger is released under the MIT license. See [LICENSE](https://github.com/genedelisa/GDLogger/blob/master/LICENSE) for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
`GDLogger` adheres to [Semantic Versioning](http://semver.org/).

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

---
## 0.0.2 - 2018-01-12
- Handle no bundle name availabile if used from a library

## 0.0.1 - 2017-12-19
### Added
- This CHANGELOG file
- Added aggregate target for Documentation that uses jazzy
- Added Swiftlint checking
- Reconfigured inits
- Updated readme.md

## [Master](https://github.com/genedelisa/GDLogger)
### Added

### Changed

### Removed


Tests/GDLoggerSpec.swift
//
//  GDLoggerSpec.swift
//  GDLogger
//
//  Created by Gene De Lisa on 04/10/16.
//  Copyright © 2017 genedelisa. All rights reserved.
//

import Quick
import Nimble
@testable import GDLogger

class GDLoggerSpec: QuickSpec {

    override func spec() {

        describe("GDLoggerSpec") {
            
            it("work by default") {
                let log = GDLogger()
                log.debug("hello")
                expect(log.logger).toNot(beNil())
            }
            
            it("has a category") {
                let log = GDLogger(category: "some category")
                log.debug("hello")
                expect(log.logger).toNot(beNil())
            }
            
            it("has a string subsystem and category") {
                let log = GDLogger("foosubsystem", category: "foocategory")
                log.debug("hello")
                expect(log.logger).toNot(beNil())
            }
            
        }

    }

}


Sources/Core/GDLogger.swift
//
//  GDLogger.swift
//  GDLogger
//
//  Created by Gene De Lisa on 02/10/17.
//  Copyright © 2017 genedelisa. All rights reserved.
//

import Foundation
import os.log


/// This is a façade for Apple's Unified Logging `OSLog`.
/// It falls back to `NSLog` in some cases.
public struct GDLogger {
    
    /// the system logger
    public var logger: OSLog?
    
    /// Predefined subsystems.
    public enum LogSubsystem: String {
        case `default`
        case bundle
        case core
        case db
        case net
        case media
        case io
        case model
        case view
        case controller
    }
    
    /// The Category.
    public enum LogCategory: String {
        case general
        case params
        case paramsController
    }
    
    /// emitted before message in debug level
    public var debugPrefix = "😺😺😺 "
    /// emitted after message in debug level
    public var debugPostfix = " 😺😺😺"
    
    /// emitted before message in error level
    public var errorPrefix = "‼️😿‼️ "
    /// emitted after message in error level
    public var errorPostfix = " ‼️😿‼️"
    
    /// emitted before message in info level
    public var infoPrefix = "ℹ️😼ℹ️ "
    /// emitted after message in info level
    public var infoPostfix = " ℹ️😼ℹ️"
    
    /// emitted before message in fault level
    public var faultPrefix = "🙀🙀‼️ "
    /// emitted after message in fault level
    public var faultPostfix = " 🙀🙀‼️"
    
    /// message format separator
    public var msgSeparator = " ☞ "
    public var fileSeparator = " 🗄"
    public var lineSeparator = "➸"
    public var columnSeparator = ":"
    //2017-12-18 15:54:03.611598-0500 SlowItDown[31758:2620634] [general] 😺😺😺 authorized ☞ checkMediaLibraryPermission() 🗄MediaLibraryController.swift➸121 😺😺😺
    
    
   
    
    /// Init that actually creates the `OSLog` instance. Or not.
    ///
    /// - Parameters:
    ///   - subsystem: an optional subsystem. The bundle if not specified
    ///   - category: the category to use
    public init(_ subsystem: String? = nil, category: String ) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            if let subsystem = subsystem {
                logger = OSLog(subsystem: subsystem, category: category)
            } else {
                if let ss = Bundle.main.bundleIdentifier {
                    logger = OSLog(subsystem: ss, category: category)
                } else {
                    logger = OSLog(subsystem: "noBundle", category: category)
                    error("no subsystem from bundle.")
                }
            }
        } else {
            logger = nil
        }

    }
    
    /// Init with specified predefined subsystem and category
    ///
    /// - Parameters:
    ///   - subsystem: a subsystem predefined in `LogSubsystem`
    ///   - category: a category predefined in `LogCategory`
    public init(_ subsystem: LogSubsystem, category: LogCategory = .general ) {
        
        if subsystem == .bundle {
            if let ss = Bundle.main.bundleIdentifier {
                self.init(ss, category: category.rawValue)
            } else {
                self.init("noBundle", category: category.rawValue)
                error("no subsystem from bundle.")
            }
        } else {
            self.init(subsystem.rawValue, category: category.rawValue)
        }
    }

    
    /// Default init. Called with inited with no args
    ///
    /// - Parameters:
    ///   - subsystem: the bundle if not specified
    ///   - category: general if not specified
    public init(_ subsystem: String? = nil, category: LogCategory = .general ) {
        
        if let subsystem = subsystem {
            self.init(subsystem, category: category.rawValue)
        } else {
            if let ss = Bundle.main.bundleIdentifier {
                self.init(ss, category: category.rawValue)
            } else {
                self.init("noBundle", category: category.rawValue)
                error("no subsystem from bundle.")
            }
        }
    }
    
    
    /// Log a Static String with Info level.
    ///
    /// - Parameters:
    ///   - msg: a Static String. No interpolation
    ///   - args: the arguments to pass to the static string
    ///   - function: defaults to function name
    ///   - file: defaults to filename
    ///   - line: defaults to line where this was called
    ///   - column: defaults to column where this was called
    ///   - dso: dynamic shared objct. the currently loaded .dylib or .so
    /// - requires: iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0
    
    public func info(_ msg: StaticString, _ args: CVarArg, function: String = #function,
                     file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, .info, function, file, line, column, dso)
        }
    }
    
    // same as info since there is no verbose
    public func verbose(_ msg: StaticString, _ args: CVarArg, function: String = #function,
                     file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, .info, function, file, line, column, dso)
        }
    }
    
    
    public func error(_ msg: StaticString, _ args: CVarArg, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, .error, function, file, line, column, dso)
        }
    }
    
    public func debug(_ msg: StaticString, _ args: CVarArg, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, .debug, function, file, line, column, dso)
        }
    }
    
    public func fault(_ msg: StaticString, _ args: CVarArg, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, .fault, function, file, line, column, dso)
        }
        
    }
    
    
    fileprivate func logMessage(_ msg: StaticString, _ logType: OSLogType, _ function: String, _ file: String, _ line: Int32, _ column: Int32, _ dso: UnsafeRawPointer?) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            let message = format(msg, function: function, file: file, line: line, column: column)
            if let logger = logger {
                os_log("%{public}@", dso: dso, log: logger, type: logType, message)
            } else {
                os_log("%{public}@", dso: dso, type: logType, message)
            }
        }
    }
    
    
    
    // Strings as the message. Interpolation possible.
    
    fileprivate func logMessage(_ msg: String, _ prefix: String, _ postfix: String, _ logType: OSLogType, _ function: String, _ file: String, _ line: Int32, _ column: Int32, _ dso: UnsafeRawPointer?) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            let message = format(msg, function: function, file: file, line: line, column: column)
            if let logger = logger {
                if let dso = dso {
                    os_log("%@%{public}@%@", dso: dso, log: logger, type: logType, prefix, message, postfix)
                } else {
                    os_log("%@%{public}@%@", log: logger, type: logType, prefix, message, postfix)
                }
            }
        } else {
            NSLog("%@", msg)
        }
    }
    
    public func debug(_ msg: String, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, debugPrefix, debugPostfix, .debug, function, file, line, column, dso)
        }
    }
    
    public func error(_ msg: String, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, errorPrefix, errorPostfix, .error, function, file, line, column, dso)
        }
        
    }
    
    public func info(_ msg: String, function: String = #function,
                     file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, infoPrefix, infoPostfix, .info, function, file, line, column, dso)
        }
        
    }
    
    // same as info since there is no "verbose"
    public func verbose(_ msg: String, function: String = #function,
                     file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, infoPrefix, infoPostfix, .info, function, file, line, column, dso)
        }
        
    }
    
    public func fault(_ msg: String, function: String = #function,
                      file: String = #file, line: Int32 = #line, column: Int32 = #column, dso: UnsafeRawPointer? = #dsohandle) {
        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOSApplicationExtension 3.0, *) {
            logMessage(msg, faultPrefix, faultPostfix, .fault, function, file, line, column, dso)
        }
    }
    
    
    fileprivate func format(_ message: String, function: String, file: String, line: Int32, column: Int32) -> String {
        let fileName = URL(fileURLWithPath: file).lastPathComponent
        return "\(message)\(msgSeparator)\(function)\(fileSeparator)\(fileName)\(lineSeparator)\(line)\(columnSeparator)\(column)"
    }
    
    fileprivate func format(_ message: StaticString, function: String, file: String, line: Int32, column: Int32) -> String {
        let fileName = URL(fileURLWithPath: file).lastPathComponent
        return "\(message)\(msgSeparator)\(function)\(fileSeparator)\(fileName)\(lineSeparator)\(line)"
    }
    
    
}
